### DISCUSSION TITLE
Revset recipes

### DISCUSSION THREAD

**Discussion by arxanas** [arxanas/git-branchless#496] [Show and tell]
git-branchless v0.4.0 supports [revset expressions](https://github.com/arxanas/git-branchless/wiki/Reference:-Revsets) for many commands. Here's a few examples of useful things you can do with revsets. Feel free to add your own recipes in the comments.

### Move all other draft commits on top of this commit

```
$ git move -x ‘draft() - stack()’
```

### Hide commits generated by imerge 

As per https://github.com/arxanas/git-branchless/discussions/380.

```
$ git hide ‘draft() & message(imerge)’
```

### “Drop” commits rather than hide them

As per https://github.com/arxanas/git-branchless/discussions/48#discussioncomment-2382485.

```
$ git move -x “children($1)” -d “$1^” && git hide “$1”
```

### Insert a commit into a stack:

As per https://github.com/arxanas/git-branchless/discussions/478.

```
$ git move -I -x “$1” -d “$2”
```

### Separate all commits in this stack for a given ticket into their own stack

This is kind of like `git-branchstack`.

```
$ git move -x ‘stack() & message(JIRA-123)’ -d origin/master
```

### Move all commits touching a class of files to the top of this stack

```
$ git move -I -x ‘stack() & paths.changed("glob:*.txt")' -d 'roots(stack())'
```

### Force push all branches in the current stack

As per https://github.com/arxanas/git-branchless/discussions/45#discussioncomment-2135091.

```
$ git push --force-with-lease origin $(git query -b 'branches() & stack()')
```

Note that this can be accomplished with the `git submit` command, in `master` at the time of this writing.

### Find out about recently-upgraded dependencies, excluding Dependabot commits

```
$ git query 'paths.changed("Cargo.lock") & author.date("after:1 month ago") - author.email(dependabot)'
```

### Hide any commits in the current stack which aren’t on a branch

```
$ git hide 'stack() - ancestors(branches())'
```

### Delete any draft branches which weren’t created by me

```
$ git hide -D ‘stack(draft() & branches() - author.name(Waleed))'
```

**Comment by rouge8** [comment]
Is it possible to query for commits that aren't on any branch? I tried `git query 'draft() - branches()'` but that looks like all commits that aren't the tip of branches. For context, what I'm trying to do is hide all commits that aren't on a branch so that they can be gc'd. I came up with this alias, but I wonder if there's a way to express it as a revset: `hide-branchless = "!f() { git query --raw 'draft()' | while read commit; do if [ -z $(git branch --contains $commit) ]; then git hide $commit; fi; done; }; f"`.

**Reply by arxanas** [reply]
Yep, that should be like the "Hide any commits in the current stack which aren’t on a branch" example. You just need to write `ancestors(branches())` instead of `branches()`. (Also, you can use `:expr` as a shorthand for `ancestors(expr)`, which is more convenient for interactive use.)

**Comment by h4l** [comment]
This is a nice feature. I needed to run tests for each commit in my current WIP chain of commits to make sure none of the intermediate commits are in a broken state. Let me know if there's a more idiomatic way to do this, but I came up with this:

```
# xargs stops immediately if a command exits with 255
$ git query -r 'range(51d3d1f, HEAD)' | xargs -I{} sh -c 'git checkout {} && pytest -x || exit 255'
```

Could easily tweak the query to use this approach to run something for all draft commits rather than just the immediate ancestors.

**Reply by arxanas** [reply]
- As a shorthand, you can also write `range(51d3d1f, HEAD)` as `51d3d1f:`.
- If you're trying to just get the commits in the current stack, you can use `stack()`.
- If you're trying to get the commits in the current stack only up to the current commit, you can write `stack():` instead. (Or to exclude the current commit, you can write `stack():.^`.)

As per my comment here https://github.com/martinvonz/jj/issues/405#issuecomment-1234662218, I'd like to create a specific `git test` command for this sort of scenario. You can try using the linked `git test` command too, but I'm not sure if you can pipe in the commits listed in `git query` or not.

**Reply by h4l** [reply]
Thanks for the tips. A specific command for this would indeed be useful.

**Reply by h4l** [reply]
@arxanas Is the order of commits from `query` defined/consistent? Seems like its ordered such that a commit appears before its parent in the in the output (like git-log, or the opposite of smartlog).

I'm wondering because with a workflow where I'm linting files in a stack for issues, I want to find the first commit that introduces a problem, so that I can fix it at source. So it seems like I can just do `git query -r 'xxx' | tac` to get the commits I want with "older"/ancestor commits first.

Also, something I've found useful is to use a revset expression with smartlog to filter its output. Doing that is a good way to experiment with expressions, as it's easy to compare the filtered output with the default output to work out what your query is doing.

**Reply by claytonrcarter** [reply]
> Is the order of commits from query defined/consistent?

I asked a similar question in discord last month and [@arxanas answer](https://discord.com/channels/915309546984050709/915309546984050712/1008941564002840576) suggests that the order may be consistent but is not defined. I briefly played with adding a `sort()` fn to revsets but didn't really get anywhere and gave for lack of enthusiasm.

**Reply by h4l** [reply]
Thanks @claytonrcarter, that's useful. I guess I'll quote what @arxanas said here for anyone not in Discord (you have to join the server to view your link):
>**arxanas** — 16/08/2022 
>Mercurial has a sort function as well, so maybe. I didn't add anything like `first()` because I didn't think it was well-defined what order the commits are in; like if you do `a & b` and some `x` appears in both sets, would you expect to be in the order as if `a` were traversed first or `b`? (Some sets are optimized for " contains" checks; at the moment, we evaluate expressions in order, but a smart query planner might evaluate them in some arbitrary other order)

**Reply by arxanas** [reply]
@h4l The intermediate results of the revsets are not sorted or guaranteed to be in a particular order, but `git query` always sorts its output topologically. I've updated https://github.com/arxanas/git-branchless/wiki/Command:-git-query#use-in-scripts and opened https://github.com/arxanas/git-branchless/pull/527 to clarify this. If we added the ability for commit sets to be ordered/sorted as part of the revset language itself, then we would want to also add functionality to `git query` to preserve that ordering.

In principle, the results of evaluating the revset when passed to other functions (`git move`, `git hide`, etc.) is not defined, but it doesn't matter for any of those.

**Reply by h4l** [reply]
Perfect, thanks @arxanas!

**Comment by pokey** [comment]
Wow these are epic!  A couple quick typos in **Force push all branches in the current stack**: it uses a bad quote and is missing a closing paren.  It should be:

```
$ git push --force-with-lease origin $(git query -b 'branches() & stack()')
```

**Reply by arxanas** [reply]
Thanks for the correction. I've updated the post. By the way, you can use `git submit` on the latest master to do the same thing a little more conveniently.

**Reply by pokey** [reply]
Oh nice!  I'll have a look at that one.  The `--create` flag looks useful

**Comment by pokey** [comment]
For the git-imerge one, I think the following is a bit more precise, as your query could sometimes pick things up that end up in a branch, and misses some things as well:

```
$ git hide 'descendants(draft() & message("regex:^Merge commit|^imerge") - ancestors(branches()))'
```
